Add Quantization  Feat #2

0xrushi opened the pull request.

jasonjmcghee commented:
Thanks for the pr to this little project!

What is the tradeoff in terms of speed vs accuracy?

Also, if we're using quantized vectors for search, shouldn't we be using a quantized vector space to search?

0xrushi commented:
I couldn't find a physical difference in speed lol, maybe older laptops might observe it. 

`performSearch` uses `searchWithSql` And the queryData is getting quantized in `searchWithSql` here 
https://github.com/jasonjmcghee/portable-hnsw/blob/f90e5dea768dfb034b55d3f3805d75974ba7c01c/index.html#L193

8 bit data looks like this, so we do have an advantage of space savings.

![image](https://github.com/jasonjmcghee/portable-hnsw/assets/6279035/edcb3964-467d-4643-a85e-8014e4214365)

Thoughts?

0xrushi commented:
Edit: made it signed
![image](https://github.com/jasonjmcghee/portable-hnsw/assets/6279035/53684fe9-a620-4c91-841e-f8ae968e1de5)

jasonjmcghee commented:
I took a more thorough pass at this, and correct me if I'm misunderstanding- this seems to be 8-bit quantization, post embedding.

So what's the advantage over product quantization as suggested in #2 ? 

You can read https://www.pinecone.io/learn/series/faiss/product-quantization/ but this from my understanding treats everything as a residual instead and can reduce memory by 1-2 orders of magnitude and be (on the order of) twice as fast.

jasonjmcghee commented:
I'm trying to figure out how we can merge this, potentially as a separate / distinct approach. Maybe within a folder, so it can still be navigated to and used, but helps provide a distinct implementation.

It adds complexity and I want folks to be able to see how simple the base implementation can be.

I'm thinking product quantization should likely be a separate folder too.

What do you think?

0xrushi commented:
How about now? I moved the JS part into `scripts/`, added a new `build_index_quantized.py` and `index.html` remains unchanged (other than the `script/` path).

jasonjmcghee reviewed (commented):
Looks good, but don't you want a dedicated html? Like `/quantized.html`?

Or am I missing an implementation detail?

jasonjmcghee commented:
Also, I think you'll need to re-execute the quantized index build prices to build the new indices / link to the quantized indices

0xrushi commented:
> Looks good, but don't you want a dedicated html? Like `/quantized.html`?
> 
> Or am I missing an implementation detail?

https://github.com/jasonjmcghee/portable-hnsw/blob/e9a88f1ef1866ef2be40677226cb2b7789167ee8/index.html#L78

The script.js can be changed here.

jasonjmcghee commented:
Totally! But this is hosted so if you make the parquet files and a dedicated html you could just visit it directly! Totally cool with merging as is, but I figured you'd want to be able to demo it etc

0xrushi commented:
Separated `frankenstein/docs.parquet` and `frankenstein/docs_quantized.parquet` into distinct files

jasonjmcghee commented on index_quantized.html:
```diff
@@ -0,0 +1,82 @@
+<body>
+    <main class="container">
+        <div class="inner">
+            <h1 class="title">Use natural language to search 50k blog posts without a server</h1>
+            Or something else if you build your own index and throw in it in a CDN.
+            <a href="https://github.com/jasonjmcghee/portable-hnsw">What is going on here? / Show me the code</a>
+            <i>Too slow for you? I'm probably doing things poorly. PRs welcome!</i>
+            <div class="path-input">
+              <a href="" id="frankenstein">Frankenstein (fastest)</a>
+              <a href="" id="shakespeare">Shakespeare</a>
+              <a href="" id="blogs">50k Blogs Dataset (slowest - ~15 seconds)</a>
+            </div>
+            <div class="path-input">
+                Path: <input type="text" id="path" placeholder="https://jasonjmcghee.github.io/portable-hnsw/frankenstein" value="frankenstein"></input>
+            </div>
+            <div class="path-input">Debounced auto-search<input id="instant-mode" type="checkbox"></input></div>
+            <div>
+                <div class="query">
+                    <textarea id="text-input">falling in love</textarea>
+                    <button disabled id="search">Search</button>
+                </div>
```
not trying to be pedantic here, would love to get this merged in! But at the moment this won't work properly according to the instructions here. It will **only** work with quantized indices.

jasonjmcghee replied:
I can update the readme and / or this text if you'd like to describe what's going on. lmk!

0xrushi replied:
looks good now?

0xrushi replied:
updated, looks good now?

jasonjmcghee replied:
A month went by - sorry to not engage more, but AFAICT this still won't work properly unless i'm misunderstanding things.

If I'm reading things correctly, it's referencing datasets in this git repo (shakespeare, frakenstein, blogs 50k) that are not quantized. So I believe this will always fail, as they are referencing the original datasets, not quantized versions.

0xrushi replied:
hey, hehe no worries I love to brainstorm!, I had almost forgotten what did I write lol. Could you point where is this?

If its the html files, the dataset remains the same, it the query that gets quantized and then is compared with the quantized `.parquet` files from `vevctordb_utils_quantized.js`
